### ISSUE TITLE
2.2dev3 incorrect url rewrite following a 302

### ISSUE THREAD

**Issue by wdauchy** [haproxy/haproxy#521]
## Output of `haproxy -vv` and `uname -a`

```
HA-Proxy version 2.2-dev3 2020/02/25 - https://haproxy.org/
Status: development branch - not safe for use in production.
Known bugs: https://github.com/haproxy/haproxy/issues?q=is:issue+is:open
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = gcc
  CFLAGS  = -O2 -g -Wdeclaration-after-statement -fwrapv -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-old-style-declaration -Wno-ignored-qualifiers -Wno-clobbered -Wno-missing-field-initializers -Wtype-limits -DTCP_USER_TIMEOUT=18
  OPTIONS = USE_PCRE=1 USE_PCRE_JIT=1 USE_LINUX_TPROXY=1 USE_GETADDRINFO=1 USE_OPENSSL=1 USE_ZLIB=1 USE_TFO=1 USE_NS=1 USE_SYSTEMD=1

Feature list : +EPOLL -KQUEUE -MY_EPOLL -MY_SPLICE +NETFILTER +PCRE +PCRE_JIT -PCRE2 -PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H -VSYSCALL +GETADDRINFO +OPENSSL -LUA +FUTEX +ACCEPT4 -MY_ACCEPT4 +ZLIB -SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL +SYSTEMD -OBSOLETE_LINKER +PRCTL +THREAD_DUMP -EVPORTS

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=32).
Built with OpenSSL version : OpenSSL 1.1.1d  10 Sep 2019
Running on OpenSSL version : OpenSSL 1.1.1d  10 Sep 2019
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with network namespace support.
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE version : 8.32 2012-11-30
Running on PCRE version : 8.32 2012-11-30
PCRE library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with zlib version : 1.2.7
Running on zlib version : 1.2.7
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with the Prometheus exporter as a service

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTTP       side=FE|BE     mux=H2
            fcgi : mode=HTTP       side=BE        mux=FCGI
       <default> : mode=HTTP       side=FE|BE     mux=H1
       <default> : mode=TCP        side=FE|BE     mux=PASS

Available services :
        prometheus-exporter

Available filters :
        [SPOE] spoe
        [CACHE] cache
        [FCGI] fcgi-app
        [TRACE] trace
        [COMP] compression
```

## What's the configuration?

```
global
    log 127.0.0.1 format rfc5424 local0 info
    daemon
    stats socket /var/lib/haproxy/stats level admin mode 600 user haproxy group haproxy expose-fd listeners
    stats timeout 2m
    maxconn 524288
    user haproxy
    group haproxy
    set-dumpable
    nbproc 1
    tune.bufsize 33792
    tune.ssl.default-dh-param 2048
    ssl-default-bind-options no-sslv3 no-tlsv10 no-tlsv11 no-tls-tickets
    ssl-default-bind-ciphers ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES128-SHA256:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES128-SHA256:ECDHE-RSA-AES256-SHA384:DH+AES
    ssl-default-bind-ciphersuites TLS_AES_128_GCM_SHA256:TLS_CHACHA20_POLY1305_SHA256:TLS_AES_256_GCM_SHA384
    ssl-default-server-options no-sslv3 no-tlsv10 no-tlsv11 no-tls-tickets
    ssl-default-server-ciphers ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-AES128-SHA256:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES128-SHA256:ECDHE-RSA-AES256-SHA384:AES128-GCM-SHA256:AES256-GCM-SHA384:AES128-SHA256:AES256-SHA256:AES256-SHA
    ssl-default-server-ciphersuites TLS_AES_128_GCM_SHA256:TLS_CHACHA20_POLY1305_SHA256:TLS_AES_256_GCM_SHA384
    ssl-server-verify none
    hard-stop-after 3660s
    nbthread 16
    cpu-map auto:1/1-16 0-15
    spread-checks 5
    chroot /etc/haproxy/chroot
    strict-limits
    no busy-polling

defaults
    mode http
    log global
    option httplog
    option http-keep-alive
    option forwardfor except 127.0.0.0/8
    option redispatch 1
    option http-ignore-probes
    retries 3
    retry-on conn-failure empty-response response-timeout 0rtt-rejected
    timeout http-request 10s
    timeout queue 1s
    timeout connect 10s
    timeout client 180s
    timeout server 10s
    timeout http-keep-alive 10s
    timeout check 5s
    maxconn 524288
    balance roundrobin
    http-reuse always
    default-server inter 5s fastinter 1s fall 3 slowstart 20s observe layer7 error-limit 5 on-error fail-check pool-purge-delay 10s tfo
    http-check expect status 200 string passing
```
some http-rules setup on frontend side:
```
acl http  ssl_fc,not
    acl https ssl_fc
    http-request set-var(req.scheme) str(https) if https
    http-request set-var(req.scheme) str(http) if http

    acl dc_noenv_fqdn hdr_end(host) -i .par.foo
    http-request set-var(req.new_host) req.hdr(host),regsub(.par.foo,.par.preprod.foo,i) if dc_noenv_fqdn
    http-request redirect code 301 location %[var(req.scheme)]://%[var(req.new_host)]%[capture.req.uri] if dc_noenv_fqdn
    http-request del-header connection if { req.hdr(connection) close }
    http-request set-header x-proto ssl if https
    http-request set-header x-forwarded-proto https if https
    http-request set-header x-tls-sessionid %[ssl_fc_session_id,hex] if https
    http-request add-header x-client-ip %[src]
    http-request set-header client-ip %[src]

    http-request set-header x-real-host %[req.hdr(host)]
    http-request set-header x-server-address %[dst]
    acl content_encoding_header_exists req.hdr(content-encoding) -m found
    http-request set-header x-original-content-encoding %[req.hdr(content-encoding)] if content_encoding_header_exists
    acl host_header_exists req.hdr(host) -m found
    http-request set-header host %[req.hdr(host),field(1,:)] if host_header_exists
    acl ::disabled_http_methods method CONNECT
    http-request deny if ::disabled_http_methods
    http-request disable-l7-retry if ! METH_GET
```
also on backend side:

```
    http-request redirect scheme https code 302 if !{ ssl_fc } METH_GET
    http-request redirect scheme https code 307 if !{ ssl_fc } ! METH_GET
```

## Steps to reproduce the behavior

Use v2.2 with a http redirection including url arguments such as:

```
$ curl -s -L -X GET -D- -o/dev/null -H 'Host: foo' "http://127.0.0.1/api/rpc/server_inventory?serial=XXXX"
HTTP/1.1 302 Found
content-length: 0
location: https://foo/api/rpc/server_inventory/
cache-control: no-cache

HTTP/2 404 
server: gunicorn/20.0.4
date: Thu, 27 Feb 2020 14:40:19 GMT
content-type: text/html; charset=utf-8
vary: Origin, Cookie, Accept-Encoding
x-frame-options: SAMEORIGIN
content-length: 6503
set-cookie: foobar; expires=Thu, 25 Feb 2021 14:40:19 GMT; Max-Age=31449600; Path=/; SameSite=Lax
```

As you can see the https redirect is missing "?serial=XXXX" in url
In v2.1:

```
$ curl -s -L -X GET -D- -o/dev/null -H 'Host: foo' "http://127.0.01/api/rpc/server_inventory?serial=XXXX"
HTTP/1.1 302 Found
content-length: 0
location: https://foo/api/rpc/server_inventory?serial=XXXX
cache-control: no-cache

HTTP/2 400 
server: gunicorn/20.0.4
date: Thu, 27 Feb 2020 14:40:32 GMT
content-type: application/json
vary: Origin, Cookie
x-frame-options: SAMEORIGIN
content-length: 39
```

**Comment by wdauchy** [comment]
note: I may work on a patch but wasn't sure where to look

**Comment by TimWolla** [comment]
I wonder if this is a regression from #510?

**Comment by jmagnin** [comment]
commit c87e46881 seems to have introduced this bug.

**Comment by cognet** [comment]
This should be fixed by @jmagnin's patch, 967d3cc1057f1c2e83b7969db1406104fbf843df

**Comment by wdauchy** [comment]
thanks! tested and fixed!
